ext/dom/php_dom.c

```diff
@@ -1519,7 +1520,14 @@ static zval *dom_nodelist_read_dimension(zend_object *object, zval *offset, int
 		return NULL;
 	}
 
-	ZVAL_LONG(&offset_copy, zval_get_long(offset));
+	zend_long lval;
+	if (Z_TYPE_P(offset) == IS_LONG) {
+		ZVAL_LONG(&offset_copy, Z_LVAL_P(offset));
+	} else if (Z_TYPE_P(offset) == IS_STRING && is_numeric_string(Z_STRVAL_P(offset), Z_STRLEN_P(offset), &lval, NULL, false) == IS_LONG) {
```

Comment:
This is a pretty big behavioural change, and a BC break.
Offsets like `0.0`, `false`, `null` previously worked and would all implicitly convert to 0. Now it would result in a NULL element. Similarly, `true` would be converted to 1 but now it also always results in NULL. In particular it's dangerous that floating point numbers, that don't have a decimal part, will now break silently.

I think you need to place the numeric string check first, and only then fall back using `zval_get_long` which will do these kind of converts for you.
Perhaps an even better solution is just adding proper support for using `$a->attributes["my_attribute_name_here"]`, which is actually what the original bug report complained about. But I think this is not the expected API, so the current *idea* to check for numeric strings might be fine.

Reply:
Using the ZPP (something like ``zend_parse_parameter_long()`` in ``Zend_API.h``) API might make a lot of sense for this use case as it will already handle those weird cases. Another possible function, which is not a public API is ``zendi_try_get_long()`` defined in ``zend_operators.c``